File: ios/SpentyAI/SpentyAI/Features/Dashboard/DashboardViewModel.swift
import Foundation

@Observable
final class DashboardViewModel {

    // MARK: - State

    var summary: DashboardSummary?
    var isLoading = false
    var errorMessage = ""
    var showError = false

    // MARK: - Pending Approval State

    var pendingTransactions: [PendingTransaction] = []
    var isLoadingPending = false

    // MARK: - Sheet State

    var showNewTransaction = false
    var showAIChat = false

    // MARK: - Cash Flow Projection State

    var cashFlowProjection: CashFlowProjection?

    // MARK: - Dependencies

    private let repository: DashboardRepository
    private let emailRepository: EmailSyncRepository
    private let cashFlowRepository: CashFlowRepository

    // MARK: - Init

    init(repository: DashboardRepository = .shared, emailRepository: EmailSyncRepository = .shared, cashFlowRepository: CashFlowRepository = .shared) {
        self.repository = repository
        self.emailRepository = emailRepository
        self.cashFlowRepository = cashFlowRepository
    }

    // MARK: - Load

    @MainActor
    func loadSummary() async {
        isLoading = true
        showError = false

        do {
            summary = try await repository.getSummary()
        } catch {
            handleError(error)
        }

        isLoading = false

        // Load pending transactions and cash flow projection in background
        await loadPendingTransactions()
        await loadCashFlowProjection()
    }

    // MARK: - Refresh (pull-to-refresh)

    @MainActor
    func refresh() async {
        do {
            summary = try await repository.getSummary()
        } catch {
            handleError(error)
        }
        await loadPendingTransactions()
        await loadCashFlowProjection()
    }

    // MARK: - Pending Transactions

    @MainActor
    func loadPendingTransactions() async {
        isLoadingPending = true
        do {
            pendingTransactions = try await emailRepository.pendingReview()
        } catch {
            // Silently fail — pending section will show empty
            pendingTransactions = []
        }
        isLoadingPending = false
    }

    @MainActor
    func approvePendingTransaction(_ id: String) async {
        do {
            _ = try await emailRepository.approveTransaction(id)
            pendingTransactions.removeAll { $0.id == id }
        } catch {
            handleError(error)
        }
    }

    @MainActor
    func rejectPendingTransaction(_ id: String) async {
        do {
            _ = try await emailRepository.rejectTransaction(id)
            pendingTransactions.removeAll { $0.id == id }
        } catch {
            handleError(error)
        }
    }

    // MARK: - Cash Flow Projection

    @MainActor
    func loadCashFlowProjection() async {
        do {
            cashFlowProjection = try await cashFlowRepository.getProjection()
        } catch {
            // Non-fatal — dashboard still works without projection
            cashFlowProjection = nil
        }
    }

    /// Total next month outflow: recurring expenses + mandates + EMIs + OD interest
    var nextMonthTotalOutflow: Double {
        let expense = cashFlowProjection?.monthlyRecurringExpense ?? 0
        let mandate = cashFlowProjection?.monthlyMandateExpense ?? 0
        let emi = cashFlowProjection?.monthlyEmiTotal ?? 0
        let od = cashFlowProjection?.monthlyOdInterest ?? 0
        return expense + mandate + emi + od
    }

    var nextMonthExpense: Double {
        cashFlowProjection?.monthlyRecurringExpense ?? 0
    }

    var nextMonthMandates: Double {
        cashFlowProjection?.monthlyMandateExpense ?? 0
    }

    var nextMonthEMI: Double {
        cashFlowProjection?.monthlyEmiTotal ?? 0
    }

    var hasProjectionData: Bool {
        cashFlowProjection != nil
    }

    // MARK: - Helpers

    @MainActor
    private func handleError(_ error: Error) {
        if let apiError = error as? APIError {
            errorMessage = apiError.localizedDescription
        } else {
            errorMessage = error.localizedDescription
        }
        showError = true
    }

    // MARK: - Computed

    var netWorth: Double {
        summary?.netWorth ?? 0
    }

    var incomeThisMonth: Double {
        summary?.incomeThisMonth ?? 0
    }

    var expenseThisMonth: Double {
        summary?.expenseThisMonth ?? 0
    }

    var pendingReview: Int {
        summary?.pendingReview ?? 0
    }

    var accounts: [Account] {
        summary?.accounts ?? []
    }

    var recentTransactions: [Transaction] {
        let txns = summary?.recentTransactions ?? []
        let approved = txns.filter { ($0.status ?? "approved").lowercased() == "approved" }
        return Array(approved.prefix(10))
    }

    var hasData: Bool {
        summary != nil
    }
}

File: android-native/app/src/main/java/com/spentyai/app/features/dashboard/DashboardViewModel.kt
package com.spentyai.app.features.dashboard

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.spentyai.app.core.models.Account
import com.spentyai.app.core.models.CashFlowProjection
import com.spentyai.app.core.models.DashboardSummary
import com.spentyai.app.core.models.PendingTransaction
import com.spentyai.app.core.models.Transaction
import com.spentyai.app.core.network.ApiClient
import com.spentyai.app.core.network.ApiResult
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class DashboardUiState(
    val summary: DashboardSummary? = null,
    val isLoading: Boolean = false,
    val showError: Boolean = false,
    val errorMessage: String = "",
    val pendingTransactions: List<PendingTransaction> = emptyList(),
    val isLoadingPending: Boolean = false,
    val cashFlowProjection: CashFlowProjection? = null
)

class DashboardViewModel(private val apiClient: ApiClient) : ViewModel() {

    private val repository = DashboardRepository(apiClient)

    private val _uiState = MutableStateFlow(DashboardUiState())
    val uiState: StateFlow<DashboardUiState> = _uiState.asStateFlow()

    // --- Computed properties ---

    val netWorth: Double get() = _uiState.value.summary?.netWorth ?: 0.0
    val incomeThisMonth: Double get() = _uiState.value.summary?.incomeThisMonth ?: 0.0
    val expenseThisMonth: Double get() = _uiState.value.summary?.expenseThisMonth ?: 0.0
    val pendingReview: Int get() = _uiState.value.summary?.pendingReview ?: 0

    val accounts: List<Account> get() = _uiState.value.summary?.accounts ?: emptyList()

    val recentTransactions: List<Transaction>
        get() {
            val txns = _uiState.value.summary?.recentTransactions ?: emptyList()
            val approved = txns.filter { (it.status ?: "approved").lowercase() == "approved" }
            return approved.take(10)
        }

    val hasData: Boolean get() = _uiState.value.summary != null

    val hasProjectionData: Boolean get() = _uiState.value.cashFlowProjection != null

    val nextMonthExpense: Double
        get() {
            val projection = _uiState.value.cashFlowProjection
            val expense = projection?.monthlyRecurringExpense ?: 0.0
            val mandate = projection?.monthlyMandateExpense ?: 0.0
            return expense + mandate
        }

    val nextMonthEMI: Double
        get() = _uiState.value.cashFlowProjection?.monthlyEmiTotal ?: 0.0

    val nextMonthODInterest: Double
        get() = _uiState.value.cashFlowProjection?.monthlyOdInterest ?: 0.0

    val nextMonthTotalOutflow: Double
        get() = nextMonthExpense + nextMonthEMI + nextMonthODInterest

    // --- Actions ---

    fun loadSummary() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, showError = false) }

            when (val result = repository.getSummary()) {
                is ApiResult.Success -> {
                    _uiState.update { it.copy(summary = result.data) }
                }
                is ApiResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            showError = true,
                            errorMessage = result.error.message ?: "Failed to load dashboard"
                        )
                    }
                }
            }

            _uiState.update { it.copy(isLoading = false) }

            // Load pending and projection in background
            loadPendingTransactions()
            loadCashFlowProjection()
        }
    }

    fun refresh() {
        viewModelScope.launch {
            when (val result = repository.getSummary()) {
                is ApiResult.Success -> {
                    _uiState.update { it.copy(summary = result.data, showError = false) }
                }
                is ApiResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            showError = true,
                            errorMessage = result.error.message ?: "Failed to refresh"
                        )
                    }
                }
            }
            loadPendingTransactions()
            loadCashFlowProjection()
        }
    }

    fun dismissError() {
        _uiState.update { it.copy(showError = false) }
    }

    private suspend fun loadPendingTransactions() {
        _uiState.update { it.copy(isLoadingPending = true) }
        when (val result = repository.getPendingReview()) {
            is ApiResult.Success -> {
                _uiState.update { it.copy(pendingTransactions = result.data.transactions) }
            }
            is ApiResult.Failure -> {
                _uiState.update { it.copy(pendingTransactions = emptyList()) }
            }
        }
        _uiState.update { it.copy(isLoadingPending = false) }
    }

    private suspend fun loadCashFlowProjection() {
        when (val result = repository.getCashFlowProjection()) {
            is ApiResult.Success -> {
                _uiState.update { it.copy(cashFlowProjection = result.data) }
            }
            is ApiResult.Failure -> {
                _uiState.update { it.copy(cashFlowProjection = null) }
            }
        }
    }
}
